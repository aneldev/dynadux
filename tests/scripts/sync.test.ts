import "jest";
import {createStore} from "../../src";

interface ITodoAppState {
  todos: ITodo[];
}

interface ITodo {
  id: string;
  label: string;
  done: boolean;
}

const actions = {
  ADD_TODO: 'ADD_TODO',
  REMOVE_TODO: 'REMOVE_TODO',
};

describe('Dynadux, basics', () => {
  test('Dispatch sync actions', () => {
    let stateChanged = 0;

    const createTodoAppStore = (onChange: (state: ITodoAppState) => void) => {
      const store = createStore<ITodoAppState>({
        initialState: {
          todos: [],
        },
        onChange,
        reducers: {
          [actions.ADD_TODO]: ({state, payload}) => {
            return {
              ...state,
              todos: state.todos.concat(payload),
            };
          },
          [actions.REMOVE_TODO]: ({state, payload: todoId}) => {
            return {
              ...state,
              todos: state.todos.filter(todo => todo.id !== todoId),
            };
          },
        },
      });

      return {
        addTodo: (todo: ITodo) => store.dispatch<ITodo>(actions.ADD_TODO, todo),
        removeTodo: (todoId: string) => store.dispatch<string>(actions.REMOVE_TODO, todoId),
        getState: store.getState,
      };
    };

    const handleStateChange = (state: ITodoAppState) => {
      stateChanged++;
    };

    const todoAppStore = createTodoAppStore(handleStateChange);

    expect(todoAppStore.getState()).toMatchSnapshot('Initial state');

    todoAppStore.addTodo({id: '301', label: 'Before work beers', done: false});
    todoAppStore.addTodo({id: '302', label: 'After work beers', done: false});

    expect(todoAppStore.getState()).toMatchSnapshot('First todos');

    todoAppStore.removeTodo('302');

    expect(todoAppStore.getState()).toMatchSnapshot('After remove of 302 todo');

    expect(stateChanged).toBe(3);
  });
});
