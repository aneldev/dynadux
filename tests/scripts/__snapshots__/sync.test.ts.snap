// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynadux, basics Dispatch sync actions: After remove of 302 todo 1`] = `
Object {
  "todos": Array [
    Object {
      "done": false,
      "id": "301",
      "label": "Before work beers",
    },
  ],
}
`;

exports[`Dynadux, basics Dispatch sync actions: First todos 1`] = `
Object {
  "todos": Array [
    Object {
      "done": false,
      "id": "301",
      "label": "Before work beers",
    },
    Object {
      "done": false,
      "id": "302",
      "label": "After work beers",
    },
  ],
}
`;

exports[`Dynadux, basics Dispatch sync actions: Initial state 1`] = `
Object {
  "todos": Array [],
}
`;
